### PR TITLE
feat/TEN-31-on-behalf-of-as

### PR DESCRIPTION
**Pull request by southbite**

### CODE REVIEW

File: packages/demos/happner-2/security-exchange-as.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  const security = mesh.exchange.security;
+
+  //create our delegate user
+  await security.addUser({
+    username: 'delegate_username',
+    password: 'password',
+  });
+
+  //link to the mesh delegate
+  await security.linkGroup('_MESH_DELEGATE', 'delegate_username');
```
Comment (southbite) [inline packages/demos/happner-2/security-exchange-as.js:46]: Mesh delegate should not be allowed to delegate _ADMIN, _MESH_ADMIN - ask SMC

File: packages/happn-3/lib/services/security/service.js [hunk 2]
```diff
@@ -1539,34 +1554,48 @@ function authorizeOnBehalfOf(session, path, action, onBehalfOf, callback) {
...
+    if (e) return callback(e);
+    if (!isDelegate) {
+      return completeCall(null, false, 'session attempting to act on behalf of is not authorized');
+    }
+    this.getOnBehalfOfSession(session, onBehalfOf, session.type, (e, behalfOfSession) => {
+      if (e) return completeCall(e, false, 'failed to get on behalf of session');
+      if (!behalfOfSession) return completeCall(null, false, 'on behalf of user does not exist');
+      this.authorize(behalfOfSession, path, action, function (err, authorized, reason) {
+        if (err) return completeCall(err, false, reason);
+        if (!authorized) return completeCall(err, false, reason);
```
Comment (southbite) [inline packages/happn-3/lib/services/security/service.js:1567]: make 1566 1567 one check

File: packages/happn-3/lib/services/security/service.js [hunk 3]
```diff
@@ -1539,34 +1554,48 @@ function authorizeOnBehalfOf(session, path, action, onBehalfOf, callback) {
...
-    });
+    behalfOfSession = this.generateSession(
+      onBehalfOfUser,
+      null,
+      {
+        info: { delegatedBy: delegatedBy.user.username },
+      },
+      { session: { type: sessionType } }
+    );
     behalfOfSession.happn = delegatedBy.happn;
```
Comment (southbite) [inline packages/happn-3/lib/services/security/service.js:1595]: is this necessary - could we decouple more, could cause a leak

File: packages/happn-3/lib/services/security/users.js [hunk 4]
```diff
@@ -405,7 +416,25 @@ function __linkGroupsToUser(user, callback) {
...
 }
 
+function userBelongsToGroups(username, groupNames, callback) {
+  if (!Array.isArray(groupNames)) {
+    return callback(new Error('groupNames must be an array'));
+  }
+  this.getGroupMemberships(username, (e, memberships) => {
+    if (e) return callback(e);
+    if (groupNames.length === 0) return callback(null, false);
+    const userGroupNames = memberships.map((membership) => membership.groupName);
```
Comment (southbite) [inline packages/happn-3/lib/services/security/users.js:426]: flag for possible optimisation - so have a cache here?

File: packages/happn-3/test/unit/security/users.js [hunk 5]
```diff
@@ -882,6 +890,220 @@ require('../../__fixtures/utils/test_helper').describe({ timeout: 120000 }, func
...
+    test
+      .expect(
+        await testCorrectCacheEmptyBasedOnWhatHappened(SD_EVENTS.DELETE_USER, {
+          obj: { _meta: { path: '/_SYSTEM/_SECURITY/_USER/test_1' } },
+        })
+      )
+      .to.be(true);
+  });
+
+  it('tests userBelongsToGroups, cache', function (done) {
```
Comment (southbite) [inline packages/happn-3/test/unit/security/users.js:1055]: name these properly

File: packages/happner-2/lib/mesh.js [hunk 6]
```diff
@@ -1171,7 +1171,7 @@ Mesh.prototype._destroyElement = util.promisify(function (componentName, callbac
       },
 
       function (done) {
-        _this._mesh.elements[componentName].component.instance._detatch(_this._mesh, done);
+        _this._mesh.elements[componentName].component.instance.detatch(_this._mesh, done);
```
Comment (southbite) [inline packages/happner-2/lib/mesh.js:1174]: correct spelling detach

File: packages/happner-2/lib/mesh.js [hunk 7]
```diff
@@ -1658,27 +1662,17 @@ Mesh.prototype._createComponent = function (spec, callback) {
       );
     }
   }.bind(_this);
-
-  //TODO - we need to remove routes somewhere too:
-
-  componentInstance.initialize(
-    componentName,
-    root,
```
Comment (southbite) [inline packages/happner-2/lib/mesh.js:1666]: root taken out why?

File: packages/happner-2/lib/mesh.js [hunk 8]
```diff
@@ -2086,18 +2080,18 @@ Mesh.prototype._eachComponentDo = function (options, callback) {
...
             _this.log.debug("%s component '%s'", options.logAction, componentName);
           }
-        } catch (e) {
-          done(new Error(e));
-          return;
-        }
-        done();
-        return;
+          if (e) {
+            _this.log.error(
```
Comment (southbite) [inline packages/happner-2/lib/mesh.js:2089]: why is done(e) not called here anymore? comment this

File: packages/happner-2/lib/modules/rest/index.js [hunk 9]
```diff
@@ -254,7 +292,13 @@ Rest.prototype.__mapMethodArguments = function (
   $origin
 ) {
   const __callback = (e, response) => {
-    if (e) return this.__respond($happn, 'Call failed', null, e, res, 500);
+    if (e) {
+      if (e.message === 'unauthorized') {
+        return this.__respond($happn, e.reason, null, new Error('Access denied'), res, 403);
+      }
+      //TODO: handle malformed errors as well
```
Comment (southbite) [inline packages/happner-2/lib/modules/rest/index.js:299]: can only do when we have schemas, comment this as a 400

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

**Review by southbite — COMMENTED** [review]
30% through

File: packages/happner-cluster/test/_lib/base-config.js [hunk 10]
```diff
@@ -39,7 +39,7 @@ module.exports = function (
       services: {
         cache: {
           config: {
-            statisticsInterval: 3e3,
+            //statisticsInterval: 3e3,
```
Comment (southbite) [inline packages/happner-cluster/test/_lib/base-config.js:42]: why did I leave this commented? explain in code

File: packages/happner-cluster/test/browser-leaking-cluster/_fixtures/drop-mongo-db.js [hunk 11]
```diff
@@ -0,0 +1,18 @@
+module.exports = function (name) {
```
Comment (southbite) [inline packages/happner-cluster/test/browser-leaking-cluster/_fixtures/drop-mongo-db.js:1]: we already have one of these

File: packages/happner-cluster/test/integration/10-cluster-broker-dynamic.js [hunk 12]
```diff
@@ -544,8 +544,8 @@ require('../_lib/test-helper').describe({ timeout: 120e3 }, (test) => {
             .to.be(true);
           return stopServer(servers[1]);
         })
-        .then(() => {
-          return test.delay(3000);
+        .then(function () {
```
Comment (southbite) [inline packages/happner-cluster/test/integration/10-cluster-broker-dynamic.js:547]: this not necessary

File: packages/happner-cluster/test/integration/20-cluster-startup-dependencies.js [hunk 13]
```diff
@@ -14,42 +14,47 @@ require('../_lib/test-helper').describe({ timeout: 120e3 }, (test) => {
...
+    // start member 4
+    await cluster.member.start(helpers.configuration.construct(20, [getSeq.getNext(), 4]), 1e3);
+    // start member 5
+    await cluster.member.start(helpers.configuration.construct(20, [getSeq.getNext(), 5]), 1e3);
+    await test.delay(6e3);
     //check member 2 (depending on member 4) is accessible
     const client = await helpers.client.create(username, password, getSeq.getPort(3));
     const result = await client.exchange.component2.use();
     test.expect(result).to.be(2);
+    test.log(JSON.stringify(cluster.events.data, null, 2));
```
Comment (southbite) [inline packages/happner-cluster/test/integration/20-cluster-startup-dependencies.js:37]: do. we need to stringify when we do a test.log?

File: packages/happner-cluster/test/unit/02-broker-component.js [hunk 14]
```diff
@@ -204,6 +204,10 @@ describe('02 - unit - brokerage component', function () {
 
     var mockLogger = {
       info: () => {},
+      debug: () => {},
```
Comment (southbite) [inline packages/happner-cluster/test/unit/02-broker-component.js:207]: why we doing this twice, can we not use a function - possibly with spy?

**Review by southbite — COMMENTED** [review]
review 2022-08-11

File: packages/happner-2/lib/modules/rest/index.js [hunk 15]
```diff
@@ -60,21 +60,22 @@ Rest.prototype.login = function ($happn, req, res) {
...
       res
     );
+  }
 
   async.eachSeries(
     Object.keys(description.callMenu),
-    (accessPoint, accessPointCB) => {
-      this.__authorizeAccessPoint($happn, $origin, accessPoint, (e, authorized) => {
+    (methodURI, accessPointCB) => {
+      this.__authorizeAccessPoint($happn, $origin, methodURI, (e, authorized) => {
```
[the remark this answers is not in the document]
Comment (southbite) [reply packages/happner-2/lib/modules/rest/index.js:76]: resolved in next update

File: packages/happner-cluster/test/_lib/base-config.js (hunk 10 above)
Comment (southbite) [reply packages/happner-cluster/test/_lib/base-config.js:42]: made adjustable in next commit

File: packages/happner-2/lib/modules/rest/index.js [hunk 16]
```diff
@@ -60,21 +60,22 @@ Rest.prototype.login = function ($happn, req, res) {
...
       null,
       res
     );
+  }
 
   async.eachSeries(
     Object.keys(description.callMenu),
-    (accessPoint, accessPointCB) => {
-      this.__authorizeAccessPoint($happn, $origin, accessPoint, (e, authorized) => {
+    (methodURI, accessPointCB) => {
```
[the remark this answers is not in the document]
Comment (southbite) [reply packages/happner-2/lib/modules/rest/index.js:75]: sorted in next commit

File: packages/happner-2/lib/modules/rest/index.js [hunk 17]
```diff
@@ -141,15 +148,70 @@ Rest.prototype.__authorize = function (res, $happn, $origin, uri, successful) {
...
+    return this.__authorizeAccessPoint($happn, $origin, methodURI, (e, authorized, reason) => {
+      if (e) {
+        return callback(e);
+      }
+      if (authorized === false) {
+        // return 403 error
+        if (!reason) reason = 'Authorization failed';
+        return this.__respond($happn, reason, null, new Error('Access denied'), res, 403);
+      }
+      callback(null, _.merge($origin, { edgeAuthorized: true }));
```
[the remark this answers is not in the document]
Comment (southbite) [reply packages/happner-2/lib/modules/rest/index.js]: done - have a new test, found issue where Object.assign was polluting the onBehalfOf cache

File: packages/happner-2/lib/system/component-instance.js [hunk 18]
```diff
@@ -759,707 +132,728 @@ ComponentInstance.prototype.initialize = function (name, root, mesh, module, con
...
-  if (this.config.directResponses)
-    return _.merge(
+  #authorizeOriginMethod(methodName, origin, callback, originBindingOverride) {
+    if (origin?.edgeAuthorized === true) {
+      return callback();
+    }
+    if (
+      !this.#boundComponentInstanceFactory.originBindingNecessary(origin, originBindingOverride)
+    ) {
+      return callback(null);
```
[the remark this answers is not in the document]
Comment (southbite) [reply packages/happner-2/lib/system/component-instance.js:497]: done in latest commit

File: packages/happner-2/lib/system/component-instance.js [hunk 19]
```diff
@@ -759,707 +132,728 @@ ComponentInstance.prototype.initialize = function (name, root, mesh, module, con
...
+      this.#log.trace('component emitEvent error', e);
+    }
+  }
 
-ComponentInstance.prototype.__createSetOptions = function (originId, options) {
-  if (this.config.directResponses)
-    return _.merge(
+  #authorizeOriginMethod(methodName, origin, callback, originBindingOverride) {
+    if (origin?.edgeAuthorized === true) {
+      return callback();
```
[the remark this answers is not in the document]
Comment (southbite) [reply packages/happner-2/lib/system/component-instance.js]: done in latest commit

File: packages/happner-2/lib/modules/rest/index.js [hunk 20]
```diff
@@ -298,20 +369,29 @@ Rest.prototype.handleRequest = function (req, res, $happn, $origin) {
...
+        $happn,
+        $origin,
+        res,
+        methodURI,
+        authorizeAs,
+        (authorizedOrigin) => {
+          // will not be hit on 403
+          const callPath = methodURI.split('/');
+          //ensure we don't have a leading /
+          if (callPath.length > 4) {
```
[the remark this answers is not in the document]
Comment (southbite) [reply packages/happner-2/lib/modules/rest/index.js:385]: done but not unit tested yet

File: packages/happner-2/lib/mesh.js [hunk 21]
```diff
@@ -2015,104 +2013,117 @@ Mesh.prototype.deferStartMethod = function (componentName, component, options, c
...
 
   this._mesh.calls = this._mesh.calls || {};
   this._mesh.calls.starting = calls = {};
 
   this._eachComponent(
     options.targets,
     options.flow,
-    function (componentName, component, done) {
-      var call,
+    (componentName, component, eachComponentCallback) => {
```
[the remark this answers is not in the document]
Comment (southbite) [reply packages/happner-2/lib/mesh.js:2074]: covered

**Review by leebow — APPROVED** [review]

**Pull request merged**